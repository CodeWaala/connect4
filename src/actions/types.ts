import { DropCoinAction } from "./dropCoin";
import { ResetGameAction } from "./resetGame";
import { UpdateScoreBoard } from "./updateScoreBoard";

export type Action = DropCoinAction | ResetGameAction | UpdateScoreBoard;
