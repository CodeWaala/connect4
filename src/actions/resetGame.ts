// An action for resetting the game to an empty board.

export interface ResetGameAction {
  type: "RESET_GAME";
}

export function resetGame(): ResetGameAction {
  return {
    type: "RESET_GAME"
  };
}
