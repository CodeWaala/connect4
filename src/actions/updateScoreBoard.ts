import { Color } from "../types";

// An action for updating the score board.

export interface UpdateScoreBoard {
    type: "UPDATE_SCORE_BOARD";
    payload: {
        color: Color;
    };
}

export function updateScoreBoard(color: Color): UpdateScoreBoard {
    return {
        type: "UPDATE_SCORE_BOARD",
        payload: { color }
    };
  }