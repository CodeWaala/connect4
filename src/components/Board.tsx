import React from "react";
import cn from "classnames";
import { connect } from "react-redux";
import { RootState } from "../reducers";
import { getBoard, getCurrentPlayer, getWinner, getIsBoardFull, getScoreBoard } from "../reducers/selectors";
import { Row } from "./Row";
import { dropCoin } from "../actions/dropCoin";
import { resetGame } from '../actions/resetGame';
import { Color, ButtonColor } from "../types";
import { Button } from "./Button";
import { updateScoreBoard } from "../actions/updateScoreBoard";
import { ScoreBoard } from "./ScoreBoard";

interface Props {
  board: ReturnType<typeof getBoard>;
  color: ReturnType<typeof getCurrentPlayer>;
  winner: ReturnType<typeof getWinner>;
  dropCoin: typeof dropCoin;
  isBoardFull: ReturnType<typeof getIsBoardFull>; // tells if board is full.
  resetGame: typeof resetGame; // reset the board game.
  updateScoreBoard: typeof updateScoreBoard; // updates the score board.
  scoreBoard: ReturnType<typeof getScoreBoard>; // get the score board.
}

export class BoardComponent extends React.Component<Props> {
  dropCoin = (column: number) => () => {
    const { winner, color, dropCoin } = this.props;
    // we only allow a player to drop a coin if there is no winner yet
    if (!winner) {
      dropCoin(column, color);
    }
  };

  displayHeader() {
    // only display the winner if there is one
    if (this.props.winner) {
      return <h2>Congratulations, {this.props.winner.color} wins the game!</h2>;
    } else {
      return <h2>It's {this.props.color}'s turn to play</h2>;
    }
  }

  displayRow = (colors: Color[], key: number) => {
    return (
      <Row
        row={key}
        dropCoin={this.dropCoin}
        colors={colors}
        key={`column-${key}`}
        winner={this.props.winner}
      />
    );
  };

  getButtonTextAndColor = (): { text: string, color: ButtonColor } => {
    const { isBoardFull, winner } = this.props;

    // If the game is on-going.
    if (!isBoardFull && !winner) {
      return {
        text: 'start over',
        color: 'red',
      }
    } else if (isBoardFull || winner) { // If the game is over.
      return {
        text: 'play again',
        color: 'green',
      }
    } else {
      return {
        text: 'reset',
        color: null,
      }
    }
  }

  render() {
    const classes = cn("Game-Board");
    const { board, resetGame, scoreBoard } = this.props;
    const { text, color } = this.getButtonTextAndColor();

    return (
      <>
        {this.displayHeader()}
        <div className="Game">
          <div className={classes}>{board.map(this.displayRow)}</div>
        </div>
        <div>
          <Button id={'gameButton'} text={text} onClick={resetGame} color={color} />
        </div>
        <p>
          <ScoreBoard scores={scoreBoard} />
        </p>
      </>
    );
  }

  componentDidUpdate(prevProps: Props): void {
    const { winner, updateScoreBoard } = this.props;
    const { winner: previousWinner } = prevProps;
    if (!previousWinner && winner) { // dispatch the action, to update the score card, because we have a winner....
      const { color: winningColor  } = winner;
      updateScoreBoard(winningColor);
    }
  }
}

const mapState = (state: RootState) => ({
  board: getBoard(state),
  color: getCurrentPlayer(state),
  winner: getWinner(state),
  isBoardFull: getIsBoardFull(state),
  scoreBoard: getScoreBoard(state),
});

export const Board = connect(mapState, { dropCoin, resetGame, updateScoreBoard })(BoardComponent);
