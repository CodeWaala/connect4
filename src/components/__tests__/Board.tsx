import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Props, BoardComponent } from '../Board';
import { resetGame } from '../../actions/resetGame';


describe('Board component', () => {
    let props: Props;
    beforeEach(() => {
        props = {
            board: [
                [null, null, null, null, null, null, null],
                [null, null, null, null, null, null, null],
                [null, null, null, null, null, null, null],
                [null, null, null, null, null, null, null],
                [null, null, null, null, null, null, null],
                [null, null, null, null, null, null, null]
            ],
        color: 'red',
        winner: null,
        dropCoin: jest.fn(),
        isBoardFull: false,
        resetGame: jest.fn(),
        updateScoreBoard: jest.fn(),
        scoreBoard: { 'red': 0, 'yellow': 0}
        }
    });


    it('renders correctly initial state', () => {
        const wrapper = shallow(<BoardComponent {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('resets the game correctly', () => {
        const wrapper = mount(<BoardComponent {...props} />);
        wrapper.find('button').simulate('click');
        expect(props.resetGame).toHaveBeenCalledTimes(1);
    });
});