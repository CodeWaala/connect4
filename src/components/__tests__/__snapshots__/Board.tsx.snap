// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component renders correctly enzyme 1`] = `
<Fragment>
  <h2>
    It's 
    red
    's turn to play
  </h2>
  <div
    className="Game"
  >
    <div
      className="Game-Board"
    >
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-0"
        row={0}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-1"
        row={1}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-2"
        row={2}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-3"
        row={3}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-4"
        row={4}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-5"
        row={5}
        winner={null}
      />
    </div>
  </div>
  <div>
    <Button
      color="red"
      id="gameButton"
      onClick={[MockFunction]}
      text="start over"
    />
  </div>
  <p>
    <ScoreBoardComponent
      scores={
        Object {
          "red": 0,
          "yellow": 0,
        }
      }
    />
  </p>
</Fragment>
`;

exports[`Board component renders correctly initial state 1`] = `
<Fragment>
  <h2>
    It's 
    red
    's turn to play
  </h2>
  <div
    className="Game"
  >
    <div
      className="Game-Board"
    >
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-0"
        row={0}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-1"
        row={1}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-2"
        row={2}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-3"
        row={3}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-4"
        row={4}
        winner={null}
      />
      <Row
        colors={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        dropCoin={[Function]}
        key="column-5"
        row={5}
        winner={null}
      />
    </div>
  </div>
  <div>
    <Button
      color="red"
      id="gameButton"
      onClick={[MockFunction]}
      text="start over"
    />
  </div>
  <p>
    <ScoreBoardComponent
      scores={
        Object {
          "red": 0,
          "yellow": 0,
        }
      }
    />
  </p>
</Fragment>
`;
